lit_parrot/model.py

```diff
@@ -44,17 +49,50 @@ def _init_weights(self, module: nn.Module) -> None:
             # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L12
             module.eps = 1e-5
 
-    def forward(self, idx: torch.Tensor) -> torch.Tensor:
-        _, t = idx.size()
-        assert (
-            t <= self.config.block_size
-        ), f"Cannot forward sequence of length {t}, block size is only {self.config.block_size}"
+    def forward(
+        self, idx: torch.Tensor, max_seq_length: Optional[int] = None, input_pos: Optional[torch.Tensor] = None
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, List[KvCache]]]:
+        B, T = idx.size()
+
+        block_size = self.config.block_size
+        if max_seq_length is None:
+            max_seq_length = block_size
+        assert T <= max_seq_length, f"Cannot forward sequence of length {T}, max seq length is only {max_seq_length}"
+        assert max_seq_length <= block_size, f"Cannot attend to {max_seq_length}, block size is only {block_size}"
+        assert T <= block_size, f"Cannot forward sequence of length {T}, block size is only {block_size}"
+
+        if self.rope_cache is None:
+            self.rope_cache = self.build_rope_cache(idx)
+        if self.mask_cache is None:
+            self.mask_cache = self.build_mask_cache(idx)
+
+        cos, sin = self.rope_cache
+        if input_pos is not None:
+            cos = cos.index_select(0, input_pos)
+            sin = sin.index_select(0, input_pos)
+            mask = self.mask_cache.index_select(2, input_pos)
+            mask = mask[:, :, :max_seq_length, :max_seq_length]
+        else:
+            cos = cos[:T]
+            sin = sin[:T]
+            mask = self.mask_cache[:, :, :T, :T]
 
         # forward the model itself
         x = self.transformer.wte(idx)  # token embeddings of shape (b, t, n_embd)
 
-        for block in self.transformer.h:
-            x = block(x)
+        if input_pos is None:  # proxy for use_cache=False
+            for block in self.transformer.h:
+                x, _ = block(x, (cos, sin), mask, max_seq_length)
+        else:
+            if not self.kv_caches:
+                cache_shape = (B, self.config.n_head, max_seq_length, self.config.n_embd // self.config.n_head)
```

Comment:
What if `B` or `max_seq_length` changes for later calls to `forward` for another batch of generation inputs?

Reply:
This is a good point, maybe we can reset the cache if we notice a difference in `B` or `max_seq_length`

Reply:
That may work, but it would complicate the code, and may cause unintentional recompilation.

That's the initial concern that leading to me putting cache-kvs as inputs and outputs. It prevents input related parameters being put into the model.

Reply:
I think it's fair to have the user do `model.kv_caches.clear()` if these parameter change

Reply:
On second thought, don't we need to reset the cache everytime we have a new batch of generation inputs? If so, it wouldn't matter whether `B` and `max_seq_length` match.

Reply:
> I think it's fair to have the user do `model.kv_caches.clear()` if these parameters change

This might be the best solution, where it is up to the user to clear the cache after each batch.